### PR TITLE
 feat(doctrine): doctrine filters like laravel eloquent filters 

### LINKED ISSUE #6844: Wrong condition in ParameterExtension
**API Platform version(s) affected**: 4

**Description**  
In ParamaterExtension there is an edge case when we pass "0" in an argument value, which is ok for boolean filter for example.
see https://github.com/api-platform/core/blob/de3bed269838556d46899e78ed5603b7e6229e8b/src/Doctrine/Orm/Extension/ParameterExtension.php#L44

This condition `!($v = $parameter->getValue())` skips the parameter for further processing.

**How to reproduce**  
Use zero as an argument value.

### PR DESCRIPTION
**Pull request by vinceAmstoutz**
| Q             | A
| ------------- | ---
| Branch?      | `main`
| Tickets        | Closes #6844 
| License       | MIT
| Doc PR        | _Coming soon_

**Completes #6749.**

Goal:

Using  Dotrine filters as we use Laravel Eloquent filters. Example using  a `BooleanFilter` with Symfony and Doctrine:
```php
#[ApiResource]
#[GetCollection(
    parameters: [
        'active' => new QueryParameter(
            filter: new BooleanFilter(),
        ),

        // support also for filter aliasing
        'enabled' => new QueryParameter(
            filter: new BooleanFilter(),
            property: 'active',
        ),
    ],
)]
#[ORM\Entity]
class FilteredBooleanParameter {
    // ...
}
```

TODO:
- [x] BooleanFilter
- [x] DateFilter
- [x] RangeFilter
- [x] Numeric
- [x] Exists
- [x] OrderFilter

**Will be done in another PR**:
- SearchFilter

### CODE REVIEW

**Comment by vinceAmstoutz** [comment]
@soyuka As requested, the swagger UI doc looks like this for the `DateFilter`:
image

**Comment by vinceAmstoutz** [comment]
@soyuka As requested, the swagger UI doc looks like this for the `RangeFilter`:

image

File: src/Doctrine/Common/Filter/ManagerRegistryAwareInterface.php [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+use Doctrine\Persistence\ManagerRegistry;
+
+interface ManagerRegistryAwareInterface
+{
+    public function hasManagerRegistry(): bool;
+
+    public function getManagerRegistry(): ManagerRegistry;
+
+    public function setManagerRegistry(?ManagerRegistry $managerRegistry): void;
```
Comment (soyuka) [inline src/Doctrine/Common/Filter/ManagerRegistryAwareInterface.php:24]:
```suggestion
    public function setManagerRegistry(ManagerRegistry $managerRegistry): void;
```

File: src/Doctrine/Odm/Extension/ParameterExtension.php [hunk 2]
```diff
@@ -29,14 +34,20 @@ final class ParameterExtension implements AggregationCollectionExtensionInterfac
...
 
+    /**
+     * @throws ContainerExceptionInterface
+     * @throws NotFoundExceptionInterface
+     */
     private function applyFilter(Builder $aggregationBuilder, ?string $resourceClass = null, ?Operation $operation = null, array &$context = []): void
     {
         foreach ($operation->getParameters() ?? [] as $parameter) {
-            if (!($v = $parameter->getValue()) || $v instanceof ParameterNotFound) {
+            if (null === ($v = $parameter->getValue()) || $v instanceof ParameterNotFound) {
```
Comment (soyuka) [inline src/Doctrine/Odm/Extension/ParameterExtension.php:50]:
```suggestion
            // TODO: remove the null equality as a parameter can have a null value
            if (null === ($v = $parameter->getValue()) || $v instanceof ParameterNotFound) {
```

File: src/Doctrine/Odm/Extension/ParameterExtension.php [hunk 3]
```diff
@@ -45,14 +56,33 @@ private function applyFilter(Builder $aggregationBuilder, ?string $resourceClass
...
-                    $context['mongodb_odm_sort_fields'] = $filterContext['mongodb_odm_sort_fields'];
-                }
+            $filter = match (true) {
+                $filterId instanceof FilterInterface => $filterId,
+                \is_string($filterId) && $this->filterLocator->has($filterId) => $this->filterLocator->get($filterId),
+                default => null,
+            };
+
+            if (!($filter instanceof FilterInterface)) {
+                return;
```
Comment (soyuka) [inline src/Doctrine/Odm/Extension/ParameterExtension.php:66]:
```suggestion
                continue;
```

File: src/Doctrine/Odm/Filter/AbstractFilter.php [hunk 4]
```diff
@@ -56,18 +61,35 @@ public function apply(Builder $aggregationBuilder, string $resourceClass, ?Opera
...
     {
+        if (!$this->hasManagerRegistry()) {
+            throw new \RuntimeException('ManagerRegistry must be initialized before accessing it.');
+        }
+
         return $this->managerRegistry;
     }
 
-    protected function getProperties(): ?array
+    public function setManagerRegistry(?ManagerRegistry $managerRegistry): void
```
Comment (soyuka) [inline src/Doctrine/Odm/Filter/AbstractFilter.php:78]:
```suggestion
    public function setManagerRegistry(ManagerRegistry $managerRegistry): void
```

File: src/Doctrine/Odm/Filter/BooleanFilter.php [hunk 5]
```diff
@@ -139,4 +141,12 @@ protected function filterProperty(string $property, $value, Builder $aggregation
 
         $aggregationBuilder->match()->field($matchField)->equals($value);
     }
+
+    /**
+     * @return array<string, string>
+     */
+    public function getSchema(Parameter $parameter): array
+    {
+        return $parameter->getSchema() ?? ['type' => 'boolean'];
```
Comment (soyuka) [inline src/Doctrine/Odm/Filter/BooleanFilter.php:150]:
```suggestion
        return ['type' => 'boolean'];
```

File: src/Doctrine/Orm/Filter/AbstractFilter.php [hunk 6]
```diff
@@ -53,29 +58,43 @@ public function apply(QueryBuilder $queryBuilder, QueryNameGeneratorInterface $q
...
-    protected function getManagerRegistry(): ManagerRegistry
+    public function hasManagerRegistry(): bool
+    {
+        return $this->managerRegistry instanceof ManagerRegistry;
+    }
+
+    public function getManagerRegistry(): ManagerRegistry
     {
+        if (!$this->hasManagerRegistry()) {
+            throw new \RuntimeException('ManagerRegistry must be initialized before accessing it.');
```
Comment (soyuka) [inline src/Doctrine/Orm/Filter/AbstractFilter.php:69]:
```suggestion
            throw new RuntimeException('ManagerRegistry must be initialized before accessing it.');
```

use ApiPlatform\Metadata\RuntimeException

File: src/Doctrine/Odm/Filter/AbstractFilter.php [hunk 7]
```diff
@@ -56,18 +61,35 @@ public function apply(Builder $aggregationBuilder, string $resourceClass, ?Opera
...
-    protected function getManagerRegistry(): ManagerRegistry
+    public function hasManagerRegistry(): bool
+    {
+        return $this->managerRegistry instanceof ManagerRegistry;
+    }
+
+    public function getManagerRegistry(): ManagerRegistry
     {
+        if (!$this->hasManagerRegistry()) {
+            throw new \RuntimeException('ManagerRegistry must be initialized before accessing it.');
```
Comment (soyuka) [inline src/Doctrine/Odm/Filter/AbstractFilter.php:72]:
```suggestion
            throw new RuntimeException('ManagerRegistry must be initialized before accessing it.');
```

File: src/Doctrine/Orm/Extension/ParameterExtension.php [hunk 8]
```diff
@@ -13,14 +13,19 @@
...
+use ApiPlatform\Doctrine\Common\Filter\ManagerRegistryAwareInterface;
 use ApiPlatform\Doctrine\Common\ParameterValueExtractorTrait;
+use ApiPlatform\Doctrine\Orm\Filter\AbstractFilter;
 use ApiPlatform\Doctrine\Orm\Filter\FilterInterface;
 use ApiPlatform\Doctrine\Orm\Util\QueryNameGeneratorInterface;
 use ApiPlatform\Metadata\Exception\InvalidArgumentException;
 use ApiPlatform\Metadata\Operation;
 use ApiPlatform\State\ParameterNotFound;
 use Doctrine\ORM\QueryBuilder;
+use Psr\Container\ContainerExceptionInterface;
```
Comment (soyuka) [inline src/Doctrine/Orm/Extension/ParameterExtension.php:25]:
```suggestion
```

File: src/Doctrine/Orm/Extension/ParameterExtension.php [hunk 9]
```diff
@@ -13,14 +13,19 @@
...
 use Psr\Container\ContainerInterface;
+use Psr\Container\NotFoundExceptionInterface;
```
Comment (soyuka) [inline src/Doctrine/Orm/Extension/ParameterExtension.php:27]:
```suggestion
```

File: src/Doctrine/Orm/Extension/ParameterExtension.php [hunk 10]
```diff
@@ -31,17 +36,22 @@ final class ParameterExtension implements QueryCollectionExtensionInterface, Que
...
+        private readonly ContainerInterface $filterLocator,
+        private readonly ?ManagerRegistry $managerRegistry = null,
+    ) {
     }
 
     /**
      * @param array<string, mixed> $context
+     *
+     * @throws ContainerExceptionInterface
+     * @throws NotFoundExceptionInterface
```
Comment (soyuka) [inline src/Doctrine/Orm/Extension/ParameterExtension.php:49]:
```suggestion
```

File: src/Doctrine/Orm/Filter/BooleanFilter.php [hunk 11]
```diff
@@ -145,4 +147,12 @@ protected function filterProperty(string $property, $value, QueryBuilder $queryB
             ->andWhere(\sprintf('%s.%s = :%s', $alias, $field, $valueParameter))
             ->setParameter($valueParameter, $value);
     }
+
+    /**
+     * @return array<string, string>
+     */
+    public function getSchema(Parameter $parameter): array
+    {
+        return $parameter->getSchema() ?? ['type' => 'boolean'];
```
Comment (soyuka) [inline src/Doctrine/Orm/Filter/BooleanFilter.php:156]:
```suggestion
        return ['type' => 'boolean'];
```

File: src/Metadata/Resource/Factory/ParameterResourceMetadataCollectionFactory.php [hunk 12]
```diff
@@ -42,6 +43,7 @@ final class ParameterResourceMetadataCollectionFactory implements ResourceMetada
     public function __construct(
         private readonly PropertyNameCollectionFactoryInterface $propertyNameCollectionFactory,
         private readonly PropertyMetadataFactoryInterface $propertyMetadataFactory,
+        private readonly LoggerInterface $logger,
```
Comment (soyuka) [inline src/Metadata/Resource/Factory/ParameterResourceMetadataCollectionFactory.php:46]: as last argument + nullable, you can fix the laravel argument and put it last

File: src/Metadata/Resource/Factory/ParameterResourceMetadataCollectionFactory.php [hunk 13]
```diff
@@ -204,11 +200,42 @@ private function setDefaults(string $key, Parameter $parameter, string $resource
             $parameter = $parameter->withExtraProperties(['_query_property' => $eloquentRelation['foreign_key']] + $parameter->getExtraProperties());
         }
 
+        $parameter = $this->addFilterMetadata($parameter);
+
+        if ($filter instanceof FilterInterface) {
+            try {
+                return $this->getLegacyFilterMetadata($parameter, $operation, $filter);
+            } catch (\RuntimeException $exception) {
```
Comment (soyuka) [inline src/Metadata/Resource/Factory/ParameterResourceMetadataCollectionFactory.php:208]:
```suggestion
            } catch (RuntimeException $exception) {
```

Metadata\Exception

File: src/Metadata/Resource/Factory/ParameterResourceMetadataCollectionFactory.php [hunk 14]
```diff
@@ -204,11 +200,42 @@ private function setDefaults(string $key, Parameter $parameter, string $resource
...
+                $this->logger->alert($exception->getMessage(), ['exception' => $exception]);
```
Comment (soyuka) [inline src/Metadata/Resource/Factory/ParameterResourceMetadataCollectionFactory.php:209]:
```suggestion
                $this->logger?->alert($exception->getMessage(), ['exception' => $exception]);
```

File: src/Laravel/ApiPlatformProvider.php [hunk 15]
```diff
@@ -360,6 +360,7 @@ public function register(): void
                         new ParameterResourceMetadataCollectionFactory(
                             $this->app->make(PropertyNameCollectionFactoryInterface::class),
                             $this->app->make(PropertyMetadataFactoryInterface::class),
+                            $this->app->make(LoggerInterface::class),
```
Comment (soyuka) [inline src/Laravel/ApiPlatformProvider.php:363]: move last

File: src/Metadata/Tests/Resource/Factory/ParameterResourceMetadataCollectionFactoryTests.php [hunk 16]
```diff
@@ -22,16 +23,23 @@
...
+    /**
+     * @throws Exception
+     * @throws ResourceClassNotFoundException
+     */
     public function testParameterFactory(): void
     {
         $nameCollection = $this->createStub(PropertyNameCollectionFactoryInterface::class);
         $propertyMetadata = $this->createStub(PropertyMetadataFactoryInterface::class);
         $filterLocator = $this->createStub(ContainerInterface::class);
+        $logger = $this->createStub(LoggerInterface::class);
```
Comment (soyuka) [inline src/Metadata/Tests/Resource/Factory/ParameterResourceMetadataCollectionFactoryTests.php:42]: revert

File: src/Symfony/Bundle/Resources/config/metadata/resource.xml [hunk 17]
```diff
@@ -82,6 +82,7 @@
         <service id="api_platform.metadata.resource.metadata_collection_factory.parameter" class="ApiPlatform\Metadata\Resource\Factory\ParameterResourceMetadataCollectionFactory" decorates="api_platform.metadata.resource.metadata_collection_factory" public="false" decoration-priority="1000">
             <argument type="service" id="api_platform.metadata.property.name_collection_factory" />
             <argument type="service" id="api_platform.metadata.property.metadata_factory" />
+            <argument type="service" id="logger" />
```
Comment (soyuka) [inline src/Symfony/Bundle/Resources/config/metadata/resource.xml:85]: should be last + on-invalid=ignore

File: tests/Functional/Parameters/BooleanFilterTest.php [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+        $expectedValue = str_contains($url, '=true') || str_contains($url, '=1');
+        foreach ($filteredItems as $item) {
+            $errorMessage = \sprintf("Expected 'active' to be %s", $expectedValue ? 'true' : 'false');
+            $this->assertSame($expectedValue, $item['active'], $errorMessage);
+        }
+    }
+
+    public static function booleanFilterScenariosProvider(): \Generator
+    {
+        yield 'active_true' => ['/filtered_boolean_parameters?active=true', 2];
```
Comment (soyuka) [inline tests/Functional/Parameters/BooleanFilterTest.php:81]:
```suggestion
        yield 'active_true' => ['/filtered_boolean_parameters?active=true', 2, 'true'];
```

File: tests/Functional/Parameters/BooleanFilterTest.php [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+    {
+        $response = self::createClient()->request('GET', $url);
+        $this->assertResponseIsSuccessful();
+
+        $responseData = $response->toArray();
+        $filteredItems = $responseData['hydra:member'];
+
+        $this->assertCount($expectedCount, $filteredItems, \sprintf('Expected %d items for URL %s', $expectedCount, $url));
+
+        $expectedValue = str_contains($url, '=true') || str_contains($url, '=1');
```
Comment (soyuka) [inline tests/Functional/Parameters/BooleanFilterTest.php:72]: instead, let's add a third arg inside the data provider

File: src/Doctrine/Orm/Filter/ExistsFilter.php [hunk 20]
```diff
@@ -129,6 +137,12 @@ public function __construct(ManagerRegistry $managerRegistry, ?LoggerInterface $
      */
     public function apply(QueryBuilder $queryBuilder, QueryNameGeneratorInterface $queryNameGenerator, string $resourceClass, ?Operation $operation = null, array $context = []): void
     {
+        $parameter = $context['parameter'] ?? null;
+
+        if (null !== ($value = $context['filters'][$parameter?->getProperty()] ?? null)) {
+            $this->filterProperty($this->denormalizePropertyName($parameter->getProperty()), $value, $queryBuilder, $queryNameGenerator, $resourceClass, $operation, $context);
```
Comment (soyuka) [inline src/Doctrine/Orm/Filter/ExistsFilter.php:143]:
```suggestion
            $this->filterProperty($this->denormalizePropertyName($parameter->getProperty()), $value, $queryBuilder, $queryNameGenerator, $resourceClass, $operation, $context);
            return;
```

File: src/Doctrine/Orm/Filter/OrderFilter.php [hunk 21]
```diff
@@ -271,4 +271,27 @@ protected function filterProperty(string $property, $direction, QueryBuilder $qu
...
+    {
+        if (str_contains($parameter->getKey(), ':property')) {
+            $parameters = [];
+            $key = str_replace('[:property]', '', $parameter->getKey());
+            foreach (array_keys($parameter->getExtraProperties()['_properties'] ?? []) as $property) {
+                $parameters[] = new OpenApiParameter(name: \sprintf('%s[%s]', $key, $property), in: 'query');
+            }
+
+            return $parameters;
+        }
```
Comment (soyuka) [inline src/Doctrine/Orm/Filter/OrderFilter.php:293]: let's put this in a trait inside doctrine/common

PropertyPlaceholderOpenApiParameterTrait

File: src/Doctrine/Orm/Filter/OrderFilter.php [hunk 22]
```diff
@@ -271,4 +279,21 @@ protected function filterProperty(string $property, $direction, QueryBuilder $qu
...
+    public function getSchema(Parameter $parameter): array
+    {
+        if (str_contains($parameter->getKey(), ':property')) {
+            $properties = [];
+            foreach (array_keys($parameter->getExtraProperties()['_properties'] ?? []) as $property) {
+                $properties[] = [$property => ['type' => 'string', 'enum' => ['asc', 'desc']]];
+            }
+
+            return ['type' => 'object', 'properties' => $properties];
+        }
```
Comment (soyuka) [inline src/Doctrine/Orm/Filter/OrderFilter.php:295]: We shouldn't need this as PropertyPlaceholderOpenApiParameterTrait alread declares a parameter per property. If this was for Validation we should move this to the ParameterValidationResourceMetadataCollectionFactory?

File: src/Symfony/Bundle/Resources/config/metadata/resource.xml [hunk 23]
```diff
@@ -85,6 +85,7 @@
             <argument type="service" id="api_platform.metadata.resource.metadata_collection_factory.parameter.inner" />
             <argument type="service" id="api_platform.filter_locator" on-invalid="ignore" />
             <argument type="service" id="api_platform.name_converter" on-invalid="ignore" />
+            <argument type="service" id="logger" on-invalid="null" />
```
Comment (soyuka) [inline src/Symfony/Bundle/Resources/config/metadata/resource.xml:88]:
```suggestion
            <argument type="service" id="logger" on-invalid="ignore" />
```

File: src/Doctrine/Orm/Filter/ExistsFilter.php [hunk 24]
```diff
@@ -263,4 +275,27 @@ private function isAssociationNullable(AssociationMapping|array $associationMapp
...
+
+    /**
+     * @return array<string, mixed>
+     */
+    public function getSchema(Parameter $parameter): array
+    {
+        return ['type' => 'boolean'];
+    }
+
+    public function getOpenApiParameters(Parameter $parameter): OpenApiParameter|array|null
```
Comment (soyuka) [inline src/Doctrine/Orm/Filter/ExistsFilter.php:287]: we should use the trait here

File: src/Doctrine/Orm/Filter/DateFilter.php [hunk 25]
```diff
@@ -269,4 +274,25 @@ protected function addWhere(QueryBuilder $queryBuilder, QueryNameGeneratorInterf
...
+        $key = $parameter->getKey();
+
+        return [
+            new OpenApiParameter(name: $key.'[after]', in: $in),
+            new OpenApiParameter(name: $key.'[before]', in: $in),
+            new OpenApiParameter(name: $key.'[strictly_after]', in: $in),
+            new OpenApiParameter(name: $key.'[strictly_before]', in: $in),
+        ];
+    }
 }
```
Comment (soyuka) [inline src/Doctrine/Orm/Filter/DateFilter.php:298]: note to self: we should probably log if a filter doesn't support the `:property` placeholder.

**Review by soyuka — CHANGES_REQUESTED** [review]

**Comment by 1ed** [comment]
@vinceAmstoutz @soyuka I think, `composer.json` files should be updated to require the minimum version which contains the new symbols. For example for  `doctrine-orm` and `doctrine-odm` should be 

```
"api-platform/doctrine-common": "^4.1"
```

**Comment by vinceAmstoutz** [comment]
> @vinceAmstoutz @soyuka I think, `composer.json` files should be updated to require the minimum version which contains the new symbols. For example for `doctrine-orm` and `doctrine-odm` should be
> 
> ```
> "api-platform/doctrine-common": "^4.1"
> ```

Why we should do that? @1ed

**Comment by 1ed** [comment]
Because new symbols like `ManagerRegistryAwareInterface` were introduced in `api-platform/doctrine-common` and then used (in a way which is not backward compatible) in eg. `api-platform/doctrine-orm` which makes `api-platform/doctrine-orm: ^4.1` incompatible with all the versions of `api-platform/doctrine-common` before the interface was defined eg. <4.1 and IMHO this should be reflected in the package constrains, as the package is now broken with `api-platform/doctrine-common: <4.1`. And that stands for the other packages too.

**Pull request merged**